gusto/advection.py

```diff
@@ -37,74 +48,254 @@ class Advection(object, metaclass=ABCMeta):
     """
     Base class for advection schemes.
 
-    :arg state: :class:`.State` object.
-    :arg field: field to be advected
-    :arg equation: :class:`.Equation` object, specifying the equation
-    that field satisfies
-    :arg solver_parameters: solver_parameters
-    :arg limiter: :class:`.Limiter` object.
-    :arg options: :class:`.AdvectionOptions` object
+    :arg solver_parameters: (optional) solver_parameters
+    :arg limiter: (optional) :class:`.Limiter` object.
+    :arg options: (optional) :class:`.AdvectionOptions` object
     """
 
-    def __init__(self, state, field, equation=None, *, solver_parameters=None,
-                 limiter=None):
+    def __init__(self, *,
+                 solver_parameters=None,
+                 limiter=None, options=None):
 
```

Comment:
You will notice that this `__init__` method is fairly empty. It would be possible to pass in the equation here and do a lot of the manipulations that are currently in the `setup` and `_setup_from_options` method. The only thing that really has to wait is replacing the advecting velocity.